Random false positive "Missing explicit access check on entity query"

I've been hitting some issues for a month or so where randomly after switching a branch I would see several "Missing explicit access check on entity query" reported. However, they're all fine and fixed. I would move the accessCheck() call around and then run again and the error would vanish. I would put the line back to the original and it wouldn't fail anymore it would be like a ghost that's now gone.

Push forward to now, and we upgraded PHPStan and went from 1.1.20 phpstan-drupal to 1.1.25 and now we have 6 false positives in our CI which won't go away. No matter what we do with the accessCheck call they report. And if we checkout the exact same code version and run PHPStan we get zero errors. So we seem to be seeing different results depending on where it runs - somewhat similar to before where random people would get random failures and it would just disappear - but instead it's now on the CI, no longer random, and completely reproducing every time. CI is GitHub Actions hosted so ephemeral. We just can't get the PHPStan to pass anymore. Locally, we get 0 errors and cannot get them to appear. I managed to get them once but they vanished on the second run, but no luck since.

Can't rule out something really weird in our configuration but I've yet to find it. The only difference I haven't explored in detail is GitHub Actions is using Ubuntu and I am using macOS.

It's only ever been a problem with this specific error though.

OK so I have managed to reproduce on my local but only with `--debug`:

```
XXX 9:05:17 XXX$ ./vendor/bin/phpstan analyze
Note: Using configuration file XXX/phpstan.neon.
 912/912 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%



 [OK] No errors


XXX 9:05:17 XXX$ ./vendor/bin/phpstan analyze
Note: Using configuration file XXX/phpstan.neon.
 912/912 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%



 [OK] No errors


XXX 9:05:17 XXX$ ./vendor/bin/phpstan --debug analyze
Note: Using configuration file XXX/phpstan.neon.
---file-list---
 ------ ---------------------------------------------------------
  Line   XXX
 ------ ---------------------------------------------------------
  122    Missing explicit access check on entity query.
         💡 See https://www.drupal.org/node/3201242
 ------ ---------------------------------------------------------

 ------ ---------------------------------------------------
  Line   XXX
 ------ ---------------------------------------------------
  492    Missing explicit access check on entity query.
         💡 See https://www.drupal.org/node/3201242
  506    Missing explicit access check on entity query.
         💡 See https://www.drupal.org/node/3201242
  1426   Missing explicit access check on entity query.
         💡 See https://www.drupal.org/node/3201242
 ------ ---------------------------------------------------

 ------ ---------------------------------------------------------------
  Line   XXX
 ------ ---------------------------------------------------------------
  76     Missing explicit access check on entity query.
         💡 See https://www.drupal.org/node/3201242
  119    Missing explicit access check on entity query.
         💡 See https://www.drupal.org/node/3201242
 ------ ---------------------------------------------------------------

 ------ ------------------------------------------------------
  Line   XXX
 ------ ------------------------------------------------------
  534    Missing explicit access check on entity query.
         💡 See https://www.drupal.org/node/3201242
 ------ ------------------------------------------------------


 [ERROR] Found 7 errors


```

Subsequent analysis without `--debug` then reproduces, as expected due to caching. But if I remove the PHPStan cache folder and run without `--debug` there is then 0 errors, and continues to be on each run. Running again after with `--debug` the errors then appear again, presumedly as `--debug` is doing something differently (non-parallel?) and ignoring cache.

--- a/src/Type/EntityQuery/EntityQueryType.php
+++ b/src/Type/EntityQuery/EntityQueryType.php
@@ -17,14 +17,17 @@ class EntityQueryType extends ObjectType
 
     public function withAccessCheck(): self
     {
-        $type = clone $this;
+        $type = new self(
+            $this->getClassName(),
+            $this->getSubtractedType(),
+            $this->getClassReflection()
+        );
         $type->hasAccessCheck = true;
-
         return $type;
     }
 
     protected function describeAdditionalCacheKey(): string
     {
-        return $this->hasAccessCheck ? 'with-access-check' : '';
+        return $this->hasAccessCheck ? 'with-access-check' : 'without-access-check';
     }
 }

--- a/src/Type/EntityQuery/EntityQueryType.php
+++ b/src/Type/EntityQuery/EntityQueryType.php
@@ -17,7 +17,10 @@ class EntityQueryType extends ObjectType
 
     public function withAccessCheck(): self
     {
-        $type = new self(
+        // The constructor of ObjectType is under backward compatibility promise.
+        // @see https://phpstan.org/developing-extensions/backward-compatibility-promise
+        // @phpstan-ignore-next-line
+        $type = new static(
             $this->getClassName(),
             $this->getSubtractedType(),
             $this->getClassReflection()

--- a/src/Type/EntityQuery/EntityQueryType.php
+++ b/src/Type/EntityQuery/EntityQueryType.php
@@ -22,4 +22,9 @@ class EntityQueryType extends ObjectType
 
         return $type;
     }
+
+    protected function describeAdditionalCacheKey(): string
+    {
+        return $this->hasAccessCheck ? 'with-access-check' : '';
+    }
 }

--- a/tests/src/Rules/EntityQueryHasAccessCheckRuleTest.php
+++ b/tests/src/Rules/EntityQueryHasAccessCheckRuleTest.php
@@ -86,5 +86,15 @@ final class EntityQueryHasAccessCheckRuleTest extends DrupalRuleTestCase
             [__DIR__ . '/data/bug-437.php'],
             []
         ];*/
+
+        yield 'bug-475.php' => [
+            [__DIR__.'/data/bug-475.php'],
+            []
+        ];
+
+        yield 'bug-475b.php' => [
+            [__DIR__.'/data/bug-475b.php'],
+            []
+        ];
     }
 }

--- a/tests/src/Rules/data/bug-475.php
+++ b/tests/src/Rules/data/bug-475.php
@@ -1,0 +1,21 @@
+<?php
+
+/**
+ * Cache key differs outside of class so this is separate reproduction
+ *
+ * @return void
+ */
+function bug475Caching(): void
+{
+    // Here condition() return type will be cached as one that has no access check
+    \Drupal::entityQuery('node')
+        ->condition('field_test', 'foo', '=')
+        ->accessCheck(FALSE)
+        ->execute();
+
+    // Cache return on condition() will also be no access check, even though caller did, unless caller type changes cache key
+    \Drupal::entityQuery('node')
+        ->accessCheck(FALSE)
+        ->condition('field_test', 'foo', '=')
+        ->execute();
+}

--- a/tests/src/Rules/data/bug-475b.php
+++ b/tests/src/Rules/data/bug-475b.php
@@ -1,0 +1,26 @@
+<?php
+
+namespace Bug475CacheKeyExample;
+
+/**
+ * Cache key differs if in class so this tests that too
+ *
+ * @return void
+ */
+class TestClass
+{
+    public function bug475Caching(): void
+    {
+        // Here condition() return type will be cached as one that has no access check
+        \Drupal::entityQuery('node')
+            ->condition('field_test', 'foo', '=')
+            ->accessCheck(FALSE)
+            ->execute();
+
+        // Cache return on condition() will also be no access check, even though caller did, unless caller type changes cache key
+        \Drupal::entityQuery('node')
+            ->accessCheck(FALSE)
+            ->condition('field_test', 'foo', '=')
+            ->execute();
+    }
+}

